QUIC: QLOG Draft Implementation

hlandau opened the pull request:
QUIC: QLOG Draft Implementation
===============================

This is an implementation of the current QLOG draft intended to be of use to those hacking on our QUIC implementation.

An example trace loaded into qvis (namely, [multistream script 2](https://github.com/search?q=script_2[]++repo%3Aopenssl%2Fopenssl+path%3A**%2Ftest%2Fquic_multistream_test.c&type=code&ref=advsearch)):

![COVER-1](https://github.com/openssl/openssl/assets/2270010/930aa5d7-cdce-4df7-b705-9a49ec91ace5)

(For those that just want to take a quick look at what kind of output this can provide when used with [qvis](https://qvis.quictools.info/), [here is an archive of a set of logfiles produced by running `quic_multistream_test`.](https://github.com/openssl/openssl/files/12561532/LOGS.tar.gz))

Using QLOG
----------

- Apply the patches to your branch using the instructions under “Getting the patches” below.
- Build with `enable-unstable-qlog`
- Set environment variable `QLOGDIR=` to a directory where logs should be stored
- Set environment variable `QFILTER='*'` to enable all events
- Run any program which uses QUIC in the built libssl (e.g. `quic_multistream_test`)
- Log files will appear in the specified directory. The JSON-SEQ format is used.
- You can load these log files into [qvis.](https://qvis.quictools.info/)

For details, see:

- [QUIC: Debugging and Tracing](https://github.com/hlandau/openssl/blob/quic-qlog/doc/designs/quic-design/debugging.md)
- [QLOG Design Document](https://github.com/hlandau/openssl/blob/quic-qlog/doc/designs/quic-design/qlog.md)

The Debugging and Tracing document above also includes guidance on using packet captures and keylogs for debugging purposes, as well as on how to use QLOG.

Implementation Status
---------------------

Currently supported events:

- `connectivity:connection_started`
- `connectivity:connection_state_updated`
- `connectivity:connection_closed`
- `transport:parameters_set`
- `transport:packet_sent`
- `transport:packet_received`
- `recovery:packet_lost`

These have been chosen as having the highest immediate utility. More will be added over time.

This implementation tracks the version of QLOG supported by qvis.

PR Status
---------

Not intended for merge at this time as the format remains somewhat unstable.
Therefore, developers hacking on QUIC should focus on using this code to enhance their productivity, rather than reviewing this PR &mdash; though reviews are of course welcome.

We discussed the possibility of experimental feature branches (wherein incomplete features are hidden behind configure options and do not have any kind of public stability guarantee) at the Committers F2F and this would be helpful for maintaining features like this in the future.

Getting the patches
-------------------

This branch will be continuously maintained for the benefit of those developing QUIC. The recommended usage pattern is for developers to cherry-pick the patches onto a branch they are developing on in order to take advantage of QLOG and enhance their productivity when troubleshooting and profiling our QUIC implementation.

The patches live at `git@github.com:hlandau/openssl.git` in branch `quic-qlog`. For convenience, the base for these patches is always kept stored as a ref at branch `quic-qlog-base`. This makes for easy cherry-picking (`quic-qlog-base...quic-qlog`).

[A script called `gitqlog` has been developed](https://gist.github.com/hlandau/60a399e453be6ac552d080c1b3efe161) to automate the application of these patches to arbitrary branches, and the removal of those patches again (e.g. before publishing a PR). You do not have to use this script, but reading the script may give some inspiration as to your own preferred workflow and applicable git commands. Usage of the script is at your own risk, as git environments do vary.

Let's walk through some example use cases for `gitqlog`:

- **Applying.** You want to apply the current published version of the patches to your current development branch.

  You run `gitqlog apply`.

  This is equivalent to:

  1. Making sure you don't already have the QLOG commits applied:

     ```shell
     git log --oneline -n 1000 | grep '^QLOG: '
     ```

  2. Fetching the commits into a dummy branch (overwriting any existing ref with the same name):

     ```shell
     $ git fetch git@github.com:hlandau/openssl.git \
        +quic-qlog:TEMP-quic-qlog +quic-qlog-base:TEMP-quic-qlog-base
     ```

  3. Cherry picking the commits onto the current branch:

     ```shell
     $ git cherry-pick --keep-redundant-commits \
        TEMP-quic-qlog-base...TEMP-quic-qlog
     ```

   `gitqlog apply` does this all for you and does the safety check to ensure you don't already have the commits applied for you.

- **Dropping.** You want to remove the QLOG patches from your branch.

  You run `gitqlog drop`.

  This is equivalent to:

  1. Running `git rebase -i ${COMMIT}^`, where `$COMMIT` is the oldest commit the commit message of which begins with `QLOG: `, and changing all lines in that file which begin with `pick ... QLOG: ` to start with `drop`.

- **Updating.** You already have the QLOG patches applied and want to update them to the latest version.

  Run `gitqlog drop` followed by `gitqlog apply`.

Composition of the PR
-------------------------

For those that do choose to review this PR, this PR is intended to be reviewable on a commit-by-commit basis.

*All* commits in this PR will always have a commit message starting with `QLOG: ` to make it easy for people to apply and then remove these commits from their own branches. The only exception are some trivial fixes discovered during development of this functionality which will also be raised in a separate PR imminently.

```
6336a407e7 (self/quic-qlog) QLOG: Add debugging guide
743035f9fe QLOG: Events: Implement transport:packet_received
b319709767 QLOG: Events: Implement transport:packet_sent
f35276878f QLOG: Wiring: QUIC QTX
1c57112b58 QLOG: QUIC MULTISTREAM TEST: Pass script name to QLOG
e1098334d8 QLOG: Wire title-setting code to QUIC_CHANNEL and SSL_CTX
7227b5fe79 QLOG: QUIC CHANNEL: Allow a log title to be specified
2a1472a06c QLOG: Events: Implement recovery:packet_lost
8d855c4eb8 QLOG: Wiring: QUIC TXPIM: Record the packet type
20375c0e25 QLOG: Wiring: QUIC FIFD
6ae8c6af16 QLOG: Wiring: QUIC QRX: Report the datagram ID from the DEMUX
5611e82f6a QLOG: Wiring: QUIC DEMUX: Report a monotonically increasing datagram ID
53ba0d4cad QLOG: Events: Implement connectivity:connection_closed
440032c448 QLOG: Events: Implement transport:parameters_set
a35ca594a2 QLOG: Events: Implement connectivity:connection_state_updated
0cfa07219a QLOG: Events: Implement connectivity:connection_started
26ce928edb QLOG: Wiring: QUIC CHANNEL
70ed273f9f QLOG: Frontend: Implementation
571105fad2 QLOG: Frontend: API Definition
4f50bbe045 QLOG: Frontend: Design
eda88cac21 QLOG: JSON Encoder: Tests
663140c3cc QLOG: JSON Encoder: Implementation
a452ea2a4a QLOG: Configuration
52b5c87b23 QLOG: JSON Encoder: Design
8793acf93b QUIC MULTISTREAM TEST: Correct trivial bug
4d0cdabc45 QUIC CHANNEL: Add missing duplicate TPARAM handling cases
```

mattcaswell commented:
Note we can't merge this until we've created the 3.2 branch.

t-j-h commented on include/internal/json_enc.h:
```diff
@@ -0,0 +1,225 @@
+    unsigned char           stack_small[16];
+    struct json_write_buf   wbuf;
+    size_t                  stack_end_byte, stack_bytes;
+} JSON_ENC;
```
I would still OSSL_ prefix these - I can see these ending up as public APIs and you named the functions with a prefix but not the typedefs and macros - make them all have a prefix.

There are so many JSON libraries out there so at least not clashing with them is prudent - and it makes promoting things later easier.

hlandau replied:
Fair enough &mdash; the reason for the naming is that symbols do actually have to have the `ossl_` prefix (e.g. to avoid conflicts in statically linked libraries) whereas the same isn't the case for typedefs.

I'll make the change.

t-j-h commented on include/internal/json_enc.h:
```diff
@@ -0,0 +1,225 @@
+ * If JSON_FLAG_IJSON is set, integers outside the range `[-2**53 + 1, 2**53 -
+ * 1]` are automatically converted to decimal strings before serialization.
+ */
+#define JSON_FLAG_NONE    0
```
OSSL_ prefix

t-j-h commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,702 @@
+        p[bit_no / BITS_PER_WORD] &= mask;
+}
+
+/* TODO abort */
```
???

hlandau replied:
Heh, this isn't about abort() if you're curious &mdash; at one point I was thinking we may need the ability to log entries transactionally and roll back if an error occurs during the process of logging an entry. Currently however the actual logging code is simple enough that this doesn't prove necessary.

Anyway, I forgot to remove this comment; removed.

t-j-h commented:
This is *nice* and would be good to see in the mainline code. Just a couple of nits noted in reviewing.

hlandau commented:
@t-j-h

Of course, QLOG is still a draft. Strictly speaking our policy probably doesn't prevent us putting this in because it's not a cryptographic algorithm (it's not like ad-hoc printf tracing output is governed by a standard either).

However I've actually become more conservative about this during the implementation of this PR as there are still a few nits and inconsistencies in the spec that I think will need to be worked out. I hope to find the time to write to the QUIC ML to get these addressed.

Again, we discussed the notion of experimental feature flags without stability guarantees at the committers F2F and this probably is an area we need to address strategically in the future. (But if we deem an `enable-unstable-qlog` flag to be adequate for now, well, I'm not going to say no...)

As the above indicates I've developed this at least for our own use for the time being. Not having this has already cost us too much productivity.

hlandau commented:
For those wanting to use the packet capture based methods discussed in the Debugging and Tracing document, this tool of mine may prove useful:

https://gist.github.com/hlandau/f67b4065909913ed31da1500dc4a8719

hlandau commented:
Perhaps we should have an "Experimental Features" technical policy that lets us carve things out from our stability guarantees. Any interest in this?

hlandau commented:
Updated to rename `JSON_ENC`.

hlandau commented:
Updated after port refactor.

hlandau commented:
@t-j-h This is already gated by a configure flag `enable-unstable-qlog`. I don't mind it going in as is subject to that. Thoughts?

t-j-h commented:
@hlandau this is not something that should require any additional actions on the part of a user to be able to enable - it is sufficiently useful and widespread used that it should be a core always-enabled capability - my view hasn't changed on this.

That the format perhaps might change over time is not an issue that should require us to leave the feature out or require a developer to jump through extra hoops to get this enabled.

t8m commented:
@hlandau does the current implementation impose any overhead if built in but not used?

hlandau commented:
> @hlandau does the current implementation impose any overhead if built in but not used?

No.

>That the format perhaps might change over time is not an issue that should require us to leave the feature out or require a developer to jump through extra hoops to get this enabled.

The format has already changed. This doesn't even implement the current draft revision because qvis doesn't implement that. So this implements a format which is *already* obsolete because the tool which is actually useful only supports an even older draft revision.

I don't understand the attitude here. "Let's ship this and randomly break it in an incompatible way later" is the diametrical opposite of every principle ordinarily underlying this project.

t8m commented:
Yeah, I actually do not understand why do we absolutely must enable it by default. Of course it is very much useful to many of our users but having to acknowledge that the format will change in future releases by configuring with `enable-unstable-qlog` for now until the RFC is published - I do not see a problem with that.

Having it as unstable and off-by-default in 3.3 of course does not mean that qlog cannot be enabled by default in a future release once the format is stable.

t8m commented:
I'd suggest merging this as is - i.e. with the off-by-default configure flag and handling whether it should be always on by default as a separate follow up OTC (or OMC?) decision.

hlandau commented:
I have no objection to merging this as-is.

hlandau commented:
Rebased and **ready for review.**

nhorman commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -71,3 +72,76 @@ void ossl_qlog_event_connectivity_connection_state_updated(QLOG *qlog,
+            /* TODO(QLOG): Consider adding ERR information in the output. */
+
+            if (m != NULL)
+                QLOG_STR("connection_code", m);
```
m can be NULL here, suggest changing the default case in qic_err_to_qlog to return "unknown" instead of NULL

hlandau replied:
That would not correspond with the schema here:

https://quicwg.org/qlog/draft-ietf-quic-qlog-quic-events.html#name-connection_closed

nhorman replied:
Nope, just missed the NULL check.  I noted it because in other locations where you pass a variable to QLOG_STR (for instance in log_packet, you call quic_pkt_type_to_qlog which has a default case that returns a string, whereas in this case you return NULL and check it.

nhorman commented on include/internal/quic_ssl.h:
```diff
@@ -127,6 +127,15 @@ QUIC_CHANNEL *ossl_quic_conn_get_channel(SSL *s);
+ * context.
+ */
+int ossl_quic_set_diag_title(SSL_CTX *ctx, const char *title);
+
```
I'm sure I'm missing something, but I don't see where this title gets used.  I would have expected it to get embedded somewhere in the json stream, but I don't

hlandau replied:
It is definitely used, see output for example:

```
{"qlog_version":"0.3","qlog_format":"JSON-SEQ","title":"quic_multistream_test: script 52","trace":{"common_fields":{"time_format":"delta","protocol_type":["QUIC"],"system_info":{"process_id":315810}},"vantage_point":{"type":"client"}}}
```

It stashes the title in the `SSL_CTX` which is then used by `port_make_channel` in `wuic_port.c`. `ossl_quic_channel_new` then copies it when instantiating QLOG.

nhorman commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -181,3 +182,432 @@ void ossl_qlog_event_recovery_packet_lost(QLOG *qlog,
+                QLOG_END()
+            QLOG_END()
+        }
+        /* TODO(QLOG): flags, length */
```
I assume its ok that this isn't implemented yet,and will be at a later date?

hlandau replied:
Yes, since these are optional fields in the schema:

https://quicwg.org/qlog/draft-ietf-quic-qlog-quic-events.html#name-packetheader

nhorman commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -181,3 +182,432 @@ void ossl_qlog_event_recovery_packet_lost(QLOG *qlog,
+                    goto cont;
+
+                rem_after = PACKET_remaining(&pkt);
+                QLOG_U64("length", rem_before - rem_after);
```
Isn't the length already encoded as part of log_frame above?

hlandau replied:
Looks like an errant line pertaining to PING only. Fixed.

Sashan reviewed (commented):
your changes look good. my comments are mostly nits/questions.

comments on long lines can be ignored. Looks like project cstyle is pretty pragmatic here. some comments are just questions, where I need to get answers to become more familiar with code here.

thanks.

hlandau commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,702 @@
+        ossl_json_object_end(&qlog->json);
+    }
+    ossl_json_object_end(&qlog->json);
+
```
It seems moot since if an error occurs output is aborted anyway. So there would be no reason to believe trying to serialize the header again later would be of any use or lead to coherent output (e.g. the header may already have been partially output and then an I/O error occurs).

hlandau commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -21,3 +21,53 @@ void ossl_qlog_event_connectivity_connection_started(QLOG *qlog,
+                                    handshake_confirmed);
+
+        if (state_s != NULL)
+            QLOG_STR("state", state_s);
```
"idle" is not a defined value:
https://quicwg.org/qlog/draft-ietf-quic-qlog-quic-events.html#section-4.6

I'm staring at this code and having difficulty seeing how QLOG_STR can be invoked with a NULL value here? It is literally inside a NULL check.

hlandau commented on ssl/quic/json_enc.c:
```diff
@@ -0,0 +1,722 @@
+            memmove(wbuf->buf,
+                    wbuf->buf + total_written,
+                    wbuf->cur - total_written);
+            wbuf->cur = 0;
```
If BIO_write_ex fails `written` must be 0 (i.e., BIO_write_ex's contract has partial writes as a success case). So if `BIO_write_ex` fails, the first byte which failed to be written must be at `wbuf->buf + total_written`? Which is then compacted by moving it to `wbuf->buf`. `wbuf->cur` indexes the next byte to write, so it becomes 0, which now contains the next unwritten byte.

hlandau commented on ssl/quic/json_enc.c:
```diff
@@ -0,0 +1,722 @@
+static void json_write_qstring(OSSL_JSON_ENC *json, const char *str);
+static void json_indent(OSSL_JSON_ENC *json);
+
+static ossl_inline int json_in_error(const OSSL_JSON_ENC *json)
```
Fixed

hlandau commented on doc/designs/quic-design/json-encoder.md:
````diff
@@ -0,0 +1,91 @@
+}
+```
+
+The zero-allocation, immediate-output design means that most API calls
````
It can happen, though generally won't. Since we have a fixed stack of 128 bits before we have to spill out to using a heap-allocated stack. This allows 128 levels of nesting composite structures which would be quite extreme for QLOG. So the JSON encoder avoids the need for allocations in likely cases, but will function correctly if it does happen.

hlandau commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,701 @@
+
+static OSSL_TIME default_now(void *arg)
+{
+    return ossl_time_now();
```
The general architectural pattern in all of our QUIC code is to ensure that the time retrieval function can be changed dynamically. This is intended to support testing. So basically any module I write has the time retrieval function as a function pointer.

It's true that QLOG has different options for encoding timestamps, but that would be handled internally by our QLOG code if we wanted to support it, so that's not what motivates this. (We use the relative encoding, see qlog_event_epilogue).

hlandau commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,701 @@
+        goto err;
+
+    if (qlog->info.now_cb == NULL)
+        qlog->info.now_cb = default_now;
```
Not needed &mdash; the `default_now` implementation doesn't use it.

hlandau commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,701 @@
+    return 1;
+}
+
+static void qlog_event_end(QLOG *qlog, int abort)
```
Ah, this is vestigial &mdash; removed.

hlandau commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,701 @@
+# undef QLOG_EVENT
+}
+
+static int lex_fail(struct lexer *lex, const char *msg)
```
Feels like we should have some way to log this at some point but not currently implemented. I'll add a TODO.

hlandau commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,627 @@
+#include "internal/quic_error.h"
+
+void ossl_qlog_event_connectivity_connection_started(QLOG *qlog,
+                                                     const QUIC_CONN_ID *init_dcid)
```
Not really actionable in any case

hlandau commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,627 @@
+#endif
+
+void ossl_qlog_event_connectivity_connection_closed(QLOG *qlog,
+                                                    const QUIC_TERMINATE_CAUSE *tcause)
```
Since our style has us indending to a parenthesis, it is not really possible to shorten this without changing the function name. For this reason it can't really be enforced as a strict rule and it's left to judgement.

hlandau commented on ssl/ssl_lib.c:
```diff
@@ -4231,6 +4231,9 @@ void SSL_CTX_free(SSL_CTX *a)
 
     OPENSSL_free(a->propq);
+#ifndef OPENSSL_NO_QLOG
+    OPENSSL_free(a->qlog_title);
```
In `ossl_quic_set_diag_title` in `ssl/quic/quic_impl.c`.

hlandau commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,627 @@
+                                     int handshake_confirmed)
+{
+    switch (state) {
+    default:
```
Would be against schema &mdash; https://quicwg.org/qlog/draft-ietf-quic-qlog-quic-events.html#name-connection_state_updated &mdash; hence this choice

Sashan reviewed (approved):
Changes look good. I think it should go in.

hlandau commented:
Added a CHANGES entry.

mattcaswell commented:
OTC: Can we confirm that we are ok to merge this PR based on a draft version of the spec?

hlandau commented:
We will have a brief discussion at OTC on if this is OK to go in.

This is still eligible for review and reviews are still sought.

mattcaswell commented on ssl/quic/json_enc.c:
```diff
@@ -0,0 +1,721 @@
+
+static void wbuf_set_bio(struct json_write_buf *wbuf, BIO *bio)
+{
+    wbuf->bio = bio;
```
Should this free any pre-existing BIO? Should this be called set0?

hlandau replied:
Renamed

mattcaswell replied:
What about my first question - should this free a pre-existing BIO?

Sashan replied:
Sorry for answering here: I think the code is OK as-is. The `wbuf->bio` instance gets freed by caller `ossl_qlog_set_sink_bio()`

mattcaswell commented on ssl/quic/json_enc.c:
```diff
@@ -0,0 +1,721 @@
+        return 0;
+
+    json->state = STATE_PRE_COMMA;
+    return ossl_json_reset(json);
```
Why is this reset necessary? Haven't we already reset everything via the memset above?

mattcaswell commented on ssl/quic/json_enc.c:
```diff
@@ -0,0 +1,721 @@
+    return wbuf_flush(&json->wbuf);
+}
+
+int ossl_json_set_sink(OSSL_JSON_ENC *json, BIO *bio)
```
set0?

mattcaswell reviewed (changes requested):
I still have more review to do on this. But here are some comments so far.

hlandau commented on doc/designs/quic-design/qlog.md:
```diff
@@ -0,0 +1,120 @@
+
+- `-quic:version_information` disables a specific event type.
+
+- Partial wildcard matches are not supported at this time.
```
The example given is valid but nonsensical. Clarified the docs.

hlandau commented:
Updated

t8m commented:
**OTC: The draft version supported by qvis is OK to be supported. This PR should document the option in INSTALL.md and openssl-qlog(7) manpage. The environment variable needs to be documented in manual pages as well. The documentation must make it clear that the version supported can be updated in incompatible way in a future release.**

hlandau commented:
I thought this exchange was a valuable perspective on QLOG and how it is a little bit different from a normal standardisation process for a protocol, which seeks to form something fully rigid and interoperable, and something more fluid and pragmatic. Fundamentally QLOG is seeking to do something a bit different and it seems will always be fuzzier as a format than we necessarily are used to expecting as implementers of protocols and standards.

https://github.com/quicwg/qlog/issues/371#issuecomment-1900972178

Just something interesting to think about.

hlandau commented:
Updated to tweak CHANGES.md further, add a note to INSTALL.md, and add a manpage openssl-qlog(7).

t-j-h commented on doc/designs/quic-design/qlog.md:
```diff
@@ -0,0 +1,138 @@
+events.
+
+Writing to files or to multi-file directories, or to arbitrary BIO sinks, will
+be supported.
```
Why future tense here - this should be what this PR does now.

t-j-h commented on doc/man7/openssl-qlog.pod:
```diff
@@ -0,0 +1,248 @@
+
+=head1 DESCRIPTION
+
+OpenSSL 3.3 and later feature unstable support for generating logs in the
```
This should not be noted this way - we should have the usual History section - but nothing about Version X and later - that doesn't fit here.

e.g.

"OpenSSL includes support for generating logs in the qlog format, which ... "

And then 

History:

These functions were added in OpenSSL 3.3.

(or should it say feature/capability/option rather than "functions"

hlandau replied:
Good catch.

t-j-h commented on ssl/quic/json_enc.c:
```diff
@@ -0,0 +1,721 @@
+}
+
+/* End a JSON array. Must be matched with a call to ossl_json_array_end(). */
+void ossl_json_array_end(OSSL_JSON_ENC *json)
```
ossl_json_array_end must be matched with a call to ossl_json_array_end - that comment belongs against ossL_json_array_begin.

t-j-h commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,627 @@
+#ifndef OPENSSL_NO_QLOG
+# define MAX_ACK_RANGES 32
+
+static void ignore_res(int x) {}
```
It doesn't ...

mattcaswell commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,696 @@
+static int lex_fail(struct lexer *lex, const char *msg)
+{
+    /*
+     * TODO(QLOG): Determine how to print log messages about bad filter strings
```
What is the plan for resolving this? Do we have an issue?

hlandau replied:
What would you prefer? Unclear what our strategy should be here. Not sure we ever really just emit output because an application which links libssl is run with an environment variable set.

mattcaswell commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,633 @@
+                         (unsigned long long)tcause->error_code);
+                m = ce;
+            }
+            /* TODO(QLOG): Consider adding ERR information in the output. */
```
What is the plan to resolve this?

hlandau replied:
Not for this PR. For future. I can just mark it QLOG future. Seems overkill to create an issue for every one of these, so how about a generic "Resolve minor QLOG enhancement TODOs" issue?

mattcaswell commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,633 @@
+        /*
+         * Can't continue scanning for frames in this case as the frame length
+         * is unknown. We log the entire body the rest of the packet payload
+         * as the raw data of the frame.
```
This last sentence doesn't quite scan...is it meant to say "of the rest of"?

hlandau replied:
Fixed

mattcaswell commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,633 @@
+    uint64_t need_skip = 0;
+
+    for (i = 0; i < num_iovec; ++i) {
+        if (!PACKET_buf_init(&pkt, iovec[i].buf, iovec[i].buf_len))
```
Are iovecs guaranteed to always hold complete frames...i.e. we never split a frame across iovecs?

hlandau replied:
We do split a frame across iovecs. But a frame's *headers* are never split across iovecs. The loops on line 536 and 540 handle this correctly for everything we currently generate.

mattcaswell commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,633 @@
+                QLOG_END()
+            QLOG_END()
+        }
+        /* TODO(QLOG): flags, length */
```
What is the plan to resolve thsi?

hlandau replied:
Inessential feature which can be added in a future PR. I'll mark these QLOG FUTURE if that's convenient.

t8m replied:
Please do. Or change it to `/* Possible enhancement: handle flags, length */

mattcaswell reviewed (changes requested):
I've now finished my first pass review of this.

While I see a test of the json encoding, I don't see an actual test for qlog itself. Shouldn't we have one?

hlandau commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,696 @@
+    qlogdir_sep = ossl_determine_dirsep(qlogdir);
+
+    /* strlen("client" / "server"); strlen(".sqlog"); _; separator; _; NUL */
+    strl = l + info->odcid.id_len * 2 + 6 + 7 + 2;
```
Made this clearer

hlandau commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,696 @@
+                ossl_json_object_begin(&qlog->json);
+                {
+                    ossl_json_key(&qlog->json, "process_id");
+                    ossl_json_u64(&qlog->json, (uint64_t)getpid());
```
I was surprised to see us depend on this too. But we already depend on it in apps/lib/s_socket.c.

Probably best just to enable this for UNIX and Windows IMO.

mattcaswell replied:
But even the apps/lib/s_sockets.c instance has some windows specific hacks in it

hlandau commented:
> I've now finished my first pass review of this.
> 
> While I see a test of the json encoding, I don't see an actual test for qlog itself. >Shouldn't we have one?

Not sure about doing it as a unit test, as it seems like it would require also adding a JSON parser, which seems a bit overkill. But it would be fairly easy to test in CI. How about it?

hlandau commented:
Minor updates. Some questions in response to feedback.

t8m commented on ssl/quic/qlog_event_helpers.c:
```diff
@@ -0,0 +1,627 @@
+                                                           uint32_t old_state,
+                                                           uint32_t new_state,
+                                                           int handshake_complete,
+                                                           int handshake_confirmed)
```
The 80 chars limit is a soft limit. It cannot strictly apply in all circumstances.

t8m commented:
> Not sure about doing it as a unit test, as it seems like it would require also adding a JSON parser, which seems a bit overkill. But it would be fairly easy to test in CI. How about it?

Yes, that would be IMO sufficient.

mattcaswell commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,698 @@
+            {
+                ossl_json_key(&qlog->json, "type");
+                ossl_json_str(&qlog->json, qlog->info.is_server
+                                  ? "server" : "client");
```
Strange indentation persists here

hlandau replied:
Unsure what you're looking for here.

mattcaswell replied:
To me the indentation looks odd. I would expect the "?" to line up with the expression it relates to, i.e. the `qlog->info.is_server`

mattcaswell commented:
There are just a few outstanding comments from me left to resolve.

Also - please can you update the CI in this PR to actually enable the option in some of the builds?

hlandau commented:
- Minor updates and fixes
- Added unit test for QLOG and filter syntax
- Added CI-only integration test that confirms the multistream test is able to produce log files correctly
- Enabled qlog in CI runs

nhorman commented on test/json_test.c:
```diff
@@ -43,7 +43,7 @@ static void helper_cleanup(struct helper *h)
     h->mem_bio = NULL;
 
-    if (!h->init) {
+    if (h->init) {
```
nit: shouldn't this be if (h->init != 0)

hlandau replied:
My understanding of our conventions is that it's OK to omit this for things which are "bools", even though we don't have an explicit int type. Which is why you can write `if (!SSL_foo())` rather than `if (SSL_foo() == 0)`.

OTOH, for some variable which can admit values other than `{0, 1}`, yes, this change should be made.

nhorman commented:
approval reconfirmed

mattcaswell commented on .github/workflows/run-checker-ci.yml:
```diff
@@ -25,7 +28,7 @@ jobs:
           no-ecx,
           no-http,
-          no-legacy,
+          no-legacy enable-unstable-qlog,
```
Why in combination with no-legacy?

hlandau replied:
It was randomly chosen as an existing CI item to piggyback on, rather than adding yet *another* CI job.

Sashan reviewed (approved):
I feel my comments are NITs only here. merge with/without addressing them is fine for me.

hlandau commented on doc/designs/quic-design/qlog.md:
```diff
@@ -0,0 +1,136 @@
+the `quic:version_information` event is disabled, then all events are disabled,
+then the `quic:packet_sent` event is reenabled.
+
+Some examples of more normal filters include:
```
Hmm, it reads better to me as is.

hlandau commented:
Updated

mattcaswell commented on test/recipes/70-test_quic_multistream.t:
```diff
@@ -6,16 +6,40 @@
-plan tests => 1;
+plan tests => 2;
+
+if (!disabled('qlog') && $ENV{RUN_CI_TESTS} == "1") {
```
Is this a github environment variable? I wonder if it will catch anyone out running OpenSSL tests in their own CI?

hlandau replied:
I'll make it OSSL_RUN_CI_TESTS

mattcaswell commented:
Also, still an outstanding earlier comment that has not been responded to: https://github.com/openssl/openssl/pull/22037#discussion_r1467850340

hlandau commented on doc/man7/openssl-qlog.pod:
```diff
@@ -0,0 +1,248 @@
+
+=head1 USAGE
+
+The qlog functionality must be explicitly enabled at build time using the
```
Good point

hlandau commented on doc/man7/openssl-qlog.pod:
```diff
@@ -0,0 +1,248 @@
+
+Disable all qlog event types.
+
+=item C<* -transport:packet_received>
```
>The terms are applied in order, thus the effects of later terms override the effects of earlier terms.

It is above...

hlandau commented:
Updated.

hlandau commented on ssl/quic/qlog.c:
```diff
@@ -0,0 +1,702 @@
+    return p[bit_no / BITS_PER_WORD] & (((size_t)1) << (bit_no % BITS_PER_WORD));
+}
+
+static ossl_unused ossl_inline void bit_set(size_t *p, uint32_t bit_no, int enable)
```
This feels excessive to me ...

hlandau commented:
Needs reapproval from @nhorman

nhorman commented:
Approval confirmed

hlandau commented:
Trivial CHANGES.md conflict will be fixed by me on merge.

hlandau commented:
Merged to master. Thanks for the reviews.

gvanem commented:
I tried to play with this new feature. But linking `libssl*.dll` has some issues since none of these gets exported from `libcrypto*.dll`:
 * `ossl_safe_getenv` in `crypto/getenv.c`.
 * `ossl_ctype_check`  in `crypto/ctype.c`. 
 * `ossl_isdigit` in `crypto/ctype.c`. 
 * And even `getpid`.  It getter be `_getpid`. No dependencies on `oldnames.lib` please!

hlandau commented:
> I tried to play with this new feature. But linking `libssl*.dll` has some issues since none of these gets exported from `libcrypto*.dll`:
> 
>     * `ossl_safe_getenv` in `crypto/getenv.c`.
> 
>     * `ossl_ctype_check`  in `crypto/ctype.c`.
> 
>     * `ossl_isdigit` in `crypto/ctype.c`.
> 
>     * And even `getpid`.  It getter be `_getpid`. No dependencies on `oldnames.lib` please!

Can you open a new issue for this? Thanks.

gvanem commented:
Yes, soon.
